parser.ls

```diff
@@ -0,0 +1,210 @@
+    start = envs~push >> location
+    end = pipe envs~pop
+
+    node = (type, content) -->
+        {type, content, location: {start: locations.pop!, end: createLoc!}}
+
+    string = node 'string'
+    atom = node 'atom'
+    list = node 'list'
+
+    hasNext = -> index < length
+    current = -> source[index]
+    generate = (cond, f, ignore) ->
+        while (ignore or hasNext!) and not cond current! => f!
+
+    fail = (message, expected, found) !->
+        loc = createLoc!
+        message = "At line #{loc.line}, column #{loc.column}: #message"
+        throw new SyntaxError message, expected, found, loc
+
+    unexpectedIf = (cond) -> current >> !->
+        fail "Unexpected character: #{esc it}", void, it if cond it
+
+    expectNext = !->
+        unless hasNext!
+            env = envs.pop!
+            extra = if env? then " (#env)" else ''
+            fail "Unexpected end of source#extra", void, '<EOS>'
+
+    skipNewline = !->
+        line++
+        column := 1
+        index++ if source[index++] == '\r' and hasNext! and current! == '\n'
+
+    move = -> column++; index++
+
+    isNewline = -> current! in ['\r', '\n', '\u2028', '\u2029']
+
+    next = ->
+        expectNext!
+        if isNewline!
+            skipNewline!
+            '\n'
+        else
+            source[move!]
+
+    lookahead = ->
+        move! if res = hasNext! and source[index + 1] == it
+        res
+
+    expect = !->
+        expectNext!
+        ch = current!
+        fail "Expected #{esc it}, found #{esc ch}", it, ch if it != ch
+        move!
+
+    skipWhitespace = ->
+        while hasNext!
+            if isNewline!
+                skipNewline!
+            else if current! in [' ', '\t']
+                move!
+            else if current! == ';'
+                while hasNext! and next! != '\n' => # do nothing
+            else
+                break
+
+        hasNext!
+
+    parseQuoted = (pipe location) >> (-> it!) >> (pipe next) >> (quote) ->
+        # Duplicate the top. The same location is used for the list and atom.
+        locations.push locations[*-1]
+        list [(atom quote), parseExpr!]
+
+    wrapped = (env, left, right, type, f) -> ->
+        start env
+        expect left
+        ret = f!
+        expect right
+        end type ret
+
+    parseList = wrapped 'unclosed list', '(', ')', list, skipWhitespace >> ->
+        generate (== ')'), (parseExpr >> pipe skipWhitespace), true
+
+    # JavaScript's escape sequences for strings
+    hex = ->
+        switch next!
+            | '0' => 0 | '1' => 1 | '2' => 2 | '3' => 3 | '4' => 4
+            | '5' => 5 | '6' => 6 | '7' => 7 | '8' => 8 | '9' => 9
+            | 'A', 'a' => 10
+            | 'B', 'b' => 11
+            | 'C', 'c' => 12
+            | 'D', 'd' => 13
+            | 'E', 'e' => 14
+            | 'F', 'f' => 15
+            | _ => fail "Expected a hex digit, found #that", 'hex digit', that
+
+    parseUnicodeEscape = ->
```

Comment:
@anko 

The small amount of bit twiddling + all the logic here is basically a necessity. It is what converts the string `"ABCD"` into the hex number `0xabcd`. And in order to stringify astral characters (in the Unicode range U+10000 - U+10FFFF), I had to convert it into [two character codes](https://github.com/anko/sexpr-plus/pull/4/files#diff-0f3013ef4059ebe5cf1e6ea712680e3eR159) using [the first formula here](https://mathiasbynens.be/notes/javascript-encoding#surrogate-formulae) because of how JavaScript handles them internally (JS characters are UCS-2, but parts of the spec require UTF-8 handling, and the ES6 `\u{N}` escape returns a UTF-8 code point, possibly 2 UCS-2 characters long).

Probably closer to [heavy wizardry](http://catb.org/jargon/html/H/heavy-wizardry.html) than magic. And trust me when I say that [Acorn is worse](https://github.com/ternjs/acorn/blob/master/src/tokenize.js#L505-L525). [Esprima isn't much better WRT this](https://github.com/jquery/esprima/blob/master/src/scanner.ts#L337-L373), even though it uses plain multiplication. I'll push a cosmetic patch to try to better document this whole thing.

Here's a description of the algorithm for parsing Unicode escapes in prose:

**GetNextHex**()

When this is called, do the following:

1. Let *next* be the current character. Advance to the next character.
2. If *next* is in the range `0` to `9`, return the numeric equivalent of that character.
3. If *next* is `A` or `a`, return 10.
4. If *next* is `B` or `b`, return 11.
5. If *next* is `C` or `c`, return 12.
6. If *next* is `D` or `d`, return 13.
7. If *next* is `E` or `e`, return 14.
8. If *next* is `F` or `f`, return 15.
9. Otherwise, throw an error, as this isn't a hex digit.

**ParseUnicodeEscape**()

When this is called, do the following:

1. Assert: we've already parsed `\u`.
2. If the current character is `{`, then do the following:
  1. Advance to the next character.
  2. Let *code* be GetNextHex().
  3. Repeat until either the current character is `}` or this loop has run 5 times:
    1. Multiply *code* by 4.
    2. Let *hex* be GetNextHex().
    3. Add *hex* to *code*.
  4. Check that the current character is `}`. If not, throw an error.
  5. Advance to the next character.
3. Else, do the following:
  1. Let *hex1* be GetNextHex() * 16<sup>3</sup>.
  2. Let *hex2* be GetNextHex() * 16<sup>2</sup>.
  3. Let *hex3* be GetNextHex() * 16.
  4. Let *hex4* be GetNextHex().
  5. Let *code* be the sum of *hex1*, *hex2*, *hex3*, and *hex4*.
4. If *code* is greater than the hex value `0x10FFFF`, then throw an error, as this is too large of a code point to be represented in UTF-16.
5. If *code* is greater than the hex value `0xFFFF`, then
  1. Let *character* be the Unicode code point represented by *code*
  2. Convert *character* into a surrogate pair as defined in the Unicode standard.
  3. Return the concatenation of each surrogate resulting from the above step.
6. Else, return the Unicode code point represented by *code*

(To be honest, the only way I can document this more thoroughly than the above would be to write an entire essay on it. It requires some low-level Unicode knowledge as well as some CS background.)